### PR TITLE
Changes required to build TFRT on Windows

### PR DESCRIPTION
**Pull request by tlemo**
First set of changes required to build TFRT on Windows with the MSVC toolchain

The state with the changes in this PR:
- CPU config only for now
- Both debug & release (opt) builds work
- bef_executor & tfrt_translate can be manually executed, but running the tests require additional changes
- Additional work is required to cleanup compilation warnings

### CODE REVIEW

File: third_party/llvm/llvm.bzl [hunk 1]
```diff
@@ -305,6 +305,13 @@ win32_cmake_vars = {
 # TODO(phawkins): use a better method to select the right host triple, rather
```
Comment (chsigg) [inline third_party/llvm/llvm.bzl:305]: This file was copied from TensorFlow. Would you mind investigating whether these changes should be made there too?

File: third_party/mlir/BUILD [hunk 2]
```diff
@@ -38,6 +38,65 @@ filegroup(
     ],
 )
 
+PYBIND11_COPTS = select({
+    "@tf_runtime//:windows": [
```
Comment (chsigg) [inline third_party/mlir/BUILD:42]: Are those the same as the other two? If so, can we share them?

Presumably they would best fit in the tensorflow/toolchains repo, but tfrt does not currently depend on it. Maybe just have it in this repo somewhere with a comment where it should go eventually?

**Review by chsigg — COMMENTED** [review]
Thanks a lot Leonard, this is great!

File: .bazelrc [hunk 3]
```diff
@@ -28,19 +34,17 @@ build:nvcc --@rules_cuda//cuda:compiler=nvcc
...
-build --cxxopt=-std=c++14 --host_cxxopt=-std=c++14
-
-# Necessary linker options.
-build --linkopt=-pthread
-build --linkopt=-ldl
-build --linkopt=-lm
-build --host_linkopt=-pthread
-build --host_linkopt=-ldl
-build --host_linkopt=-lm
+build:android --cxxopt=-std=c++14
```
[the remark this answers is not in the document]
Comment (tlemo) [reply .bazelrc:38]: I agree, per-toolchain is better than per-platform. 

As a side note - today Windows is assumed to mean MSVC, but it may be worth considering a Windows build with Clang.

File: BUILD [hunk 4]
```diff
@@ -47,6 +47,21 @@ config_setting(
     visibility = ["//visibility:public"],
 )
 
+config_setting(
+    name = "windows",
+    # Internal builds query the target OS.
+    flag_values = if_google(
```
[the remark this answers is not in the document]
Comment (tlemo) [reply BUILD:53]: I tried to keep the formatting reasonable, but ultimately I imagine that the round-trip through google3 would autoformat all the files, is that not the case?

File: BUILD [hunk 5]
```diff
@@ -1159,16 +1174,22 @@ tfrt_cc_library(
         "lib/io/buffered_input_stream.cc",
         "lib/io/file_input_stream.cc",
         "lib/io/file_system.cc",
-        "lib/io/posix_file_system.cc",
-        "lib/io/posix_file_system.h",
-    ],
+    ] + select({
+        "@tf_runtime//:windows": [
```
[the remark this answers is not in the document]
Comment (tlemo) [reply BUILD:1178]: would that work ok when using TFRT in a "parent" workspace? (ex. TF) 

I see most place use the fully qualified `@tf_runtime` and I tried to do the same. If I aimed for the wrong consistency please let me know and I'll make the changes.

File: build_defs.bzl [hunk 6]
```diff
@@ -19,8 +19,71 @@
...
+
+def if_google(google_value, oss_value = []):
+    """Returns one of the arguments based on the non-configurable build env.
+
+    Specifically, it does not return a `select`, and can be used to e.g.
+    compute elements of list attributes.
+    """
+    return oss_value  # copybara:comment_replace return google_value
+
 TFRT_COPTS = select({
```
[the remark this answers is not in the document]
Comment (tlemo) [reply build_defs.bzl:38]: I'd be happy to, but:
1. TFRT_COPTS (and other TFRT_xxx variables) preceded my changes, would you like to change them all?
2. I'm considering reusing TFRT_COPTS to factor out common flags and use it in other build files in a follow up iteration

What do you think?

File: include/tfrt/bef/bef_reader.h [hunk 7]
```diff
@@ -128,14 +128,14 @@ class BEFKernel {
   // num_used_bys as the number of results is not fixed.) in BEF. kernel_code,
   // kernel_location, num_arguments, num_attributes, num_functions, and
   // num_results in BEF can be directly mapped using this struct.
-  struct __attribute__((packed)) BEFKernelHeader {
+  PACKED_STRUCT(BEFKernelHeader {
```
[the remark this answers is not in the document]
Comment (tlemo) [reply include/tfrt/bef/bef_reader.h:131]: I belive you're right. Is the suggestion to use that explicitly and do without the PACKED_STRUCT macro?

File: include/tfrt/host_context/async_value.h [hunk 8]
```diff
@@ -349,7 +350,7 @@ class AsyncValue {
   const bool is_refcounted_ : 1;
 
   // Unused padding bits.
-  unsigned unused_ : 4;
+  // unsigned unused_ : 4;
```
[the remark this answers is not in the document]
Comment (tlemo) [reply include/tfrt/host_context/async_value.h:353]: Yes, although I should really remove the line instead of commenting it out.

Long version: the original code is not portable - bit fields packing is implementation defined, and some compilers (notably MSVC) will not pack together bitfields with different base types.

File: include/tfrt/host_context/diagnostic.h [hunk 9]
```diff
@@ -33,7 +33,8 @@ namespace tfrt {
 class ExecutionContext;
 
 // This is a simple representation of a decoded diagnostic.
-struct DecodedDiagnostic {
+class DecodedDiagnostic {
```
[the remark this answers is not in the document]
Comment (tlemo) [reply include/tfrt/host_context/diagnostic.h:36]: Yes. MSVC requires consistent struct/class usage across declarations and definitions.

This is technically a MSVC bug, but with a long history and unlikely to get fixed (the struct/class result in different name name mangling and may lead to painful to debug linking issues)

File: include/tfrt/support/aligned_buffer.h [hunk 10]
```diff
@@ -36,11 +36,18 @@ struct AlignedAllocator {
...
+
+  template <class Other>
+  constexpr AlignedAllocator(const AlignedAllocator<Other>&) noexcept {}
+
   T* allocate(size_t n) {
     auto* ptr = AlignedAlloc(Align, n * sizeof(T));
     return static_cast<T*>(ptr);
   }
-  void deallocate(T* p, size_t) noexcept { std::free(p); }
+  void deallocate(T* p, size_t) noexcept { AlignedFree(p); }
```
[the remark this answers is not in the document]
Comment (tlemo) [reply include/tfrt/support/aligned_buffer.h:50]: Good questions!

The `noexcept` was there in the original code and I left it alone, although it's not really required for conformance to the allocator interface. Since you mentioned, it's better to remove it (to avoid suboptimal code generation in this case)

Regarding the need for `AlignedFree()` vs. just calling `std::free()`: On may unix implementation, the aligned allocations can be freed with `free()`. This is not always the case, and in particular with the MSVC runtime, `_aligned_malloc` must be paired with `_aligned_free()` (trying to free aligned blocks with `free()` results in nasty runtime issues)

Finally, the additional constructors are a required part of the allocator interface. MSVC debug runtime needs these constructors for implementing some of the debug checks.

File: lib/support/alloc.cc [hunk 11]
```diff
@@ -22,21 +22,30 @@
 namespace tfrt {
 
 void* AlignedAlloc(size_t alignment, size_t size) {
-  if (alignment <= alignof(void*)) return std::malloc(size);
   size = (size + alignment - 1) / alignment * alignment;
-
-#if defined(__ANDROID__) || defined(OS_ANDROID)
+#ifdef _WIN32
+  return _aligned_malloc(size, alignment);
+#elif defined(__ANDROID__) || defined(OS_ANDROID)
```
[the remark this answers is not in the document]
Comment (tlemo) [reply lib/support/alloc.cc:28]: MSVC doesn't support `aligned_alloc()`

https://docs.microsoft.com/en-us/cpp/standard-library/cstdlib?view=msvc-160 (search for aligned_alloc)

File: third_party/llvm/llvm.bzl (hunk 1 above)
Comment (tlemo) [reply third_party/llvm/llvm.bzl:305]: I tried to compare & sync the two versions. I think the TF has everything, and I only ported over the required bits (TF has a number of extra configs which are not defined in TFRT)

File: third_party/mlir/BUILD (hunk 2 above)
Comment (tlemo) [reply third_party/mlir/BUILD:42]: Yes, this is one place where we could use TFRT_OPTS. I was thinking to keep that factoring out in a follow up PR, since there may be other changes (including potential differences in flags).

File: .bazelrc (hunk 3 above)
[the remark this answers is not in the document]
Comment (chsigg) [reply .bazelrc]: Yes, we are very much considering it. One issue is that CUDA clang on Windows is not officially supported.

File: BUILD (hunk 4 above)
[the remark this answers is not in the document]
Comment (chsigg) [reply BUILD:53]: It does format all files. The starlark formatter accepts both variants though.

```
foo = ["bar"],
```
and
```
foo = [
    "bar",
],
```

File: BUILD (hunk 5 above)
[the remark this answers is not in the document]
Comment (chsigg) [reply BUILD:1178]: It's complicated.

`@tf_runtime//...` should be used for absolute paths in .bzl macros/rules that are called from dependent repositories. That breaks though if TFRT is not imported as `tf_runtime`. We don't support that for TFRT (but we do for TF), so it's fine.

The slightly clean way is to use `Label()`. This implicitly refers to the repository it was called from.

For BUILD files or macros that are not called from dependent repositories, `//...` is fine for absolute paths. There are many unnecessary `@tf_runtime` prefixes because copybara does those rewrites on export and doesn't know when the explicit repo is necessary.

References to same-package targets really should just use `:foo`.

File: build_defs.bzl (hunk 6 above)
[the remark this answers is not in the document]
Comment (chsigg) [reply build_defs.bzl:38]: If you are going to factor them out, great. It's fine to keep them without `_` for now.

File: include/tfrt/bef/bef_reader.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (chsigg) [reply include/tfrt/bef/bef_reader.h:131]: clang 3.3 and gcc 5.4 are happy with it, that should be good enough. Yes, if possible please remove the macro.

I think we want to support gcc-8, iirc. I'm not sure about clang.

File: include/tfrt/host_context/diagnostic.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (chsigg) [reply include/tfrt/host_context/diagnostic.h:36]: Removing it seems fine, it didn't make much sense anymore after ecd68c32546434bf80f90d2ad8f32095c4cc066e anyway. ;-)

File: include/tfrt/support/aligned_buffer.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (chsigg) [reply include/tfrt/support/aligned_buffer.h]: Thanks for the explanation. Would you mind leaving some breadcrumb comments in the code about this?

File: lib/support/alloc.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (chsigg) [reply lib/support/alloc.cc:28]: I see. Can you add a link to https://developercommunity.visualstudio.com/t/c17-stdaligned-alloc%E7%BC%BA%E5%A4%B1/468021#T-N473365 in a comment?

File: third_party/llvm/llvm.bzl (hunk 1 above)
Comment (chsigg) [reply third_party/llvm/llvm.bzl:305]: Thanks!

File: third_party/mlir/BUILD (hunk 2 above)
Comment (chsigg) [reply third_party/mlir/BUILD:42]: Sounds good.

File: tools/BUILD [hunk 12]
```diff
@@ -106,7 +106,7 @@ tfrt_cc_library(
         "@tf_runtime//:basic_kernels_alwayslink",
         "@tf_runtime//:core_runtime_alwayslink",
         "@tf_runtime//:data_alwayslink",
-        "@tf_runtime//:io_alwayslink",
+        "@tf_runtime//:io",
```
Comment (chsigg) [inline tools/BUILD:109]: Why is this change needed?

**Review by chsigg — COMMENTED** [review]
LGTM from my side. I don't have time to babysit the merge at the moment though, I'm really OoO.

File: BUILD (hunk 5 above)
[the remark this answers is not in the document]
Comment (tlemo) [reply BUILD:1178]: Thanks for the explanation, fixed.

File: include/tfrt/bef/bef_reader.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (tlemo) [reply include/tfrt/bef/bef_reader.h:131]: I find the macro a bit more self descriptive, but I don't feel strongly about it. Switching to explicit #pragmas.

File: tools/BUILD (hunk 12 above)
Comment (tlemo) [reply tools/BUILD:109]: To account for the changes made to lib/io (`static_registration.cc` has been removed)

**Review by mihaimaruseac — APPROVED** [review]

**Comment by mihaimaruseac** [comment]
(approving to move forward)

File: tools/BUILD (hunk 12 above)
Comment (chsigg) [reply tools/BUILD:109]: I don't think this will quite work. The `io_alwayslink` target uses `alwayslink = True`, which is required for the static registration but we also want to have the normal `io` without registration.

Would it work to have the posix and windows registration functions use the same name (exactly one of them is included in the srcs), and then call that from static_registration.cc?

I'm getting back to you on this because changing this particular line breaks the copybara import: there are google internal targets just below that are stripped/inserted during export/import, and the way this is setup prevents nearby changes. This is something we will need to fix on our side, but here I believe this line should stay unchanged.

Comment (tlemo) [reply tools/BUILD:109]: I was afraid this may be the case. Putting things back into `static_registration.cc` is not too bad though, it should just require a platform `#ifdef`. I'll try to get these changes ready asap.

Comment (tlemo) [reply tools/BUILD]: Done. 

> Would it work to have the posix and windows registration functions use the same name (exactly one of the two definitions is included in the srcs), and then call that from static_registration.cc?

This idea worked out nicely, thanks.

**Comment by chsigg** [comment]
Hi Leonard. I'm in the process of manually merging most of this PR, but it might not land until tomorrow. Just fyi. Thanks!

**Comment by tlemo** [comment]
> ... merging most of this PR

Are you expecting any parts which can't / won't be merged?

**Comment by chsigg** [comment]
Hi Leonard, the change is partially merged now. I kept some things out that I wasn't quite sure if they are needed. I thought it's best to get everything else in first and then we can take it from there. It doesn't mean that I object to the other changes.

I will be out of office until next week, but happy to review more on Monday.

Thanks a lot for all the hard work!

**Comment by tlemo** [comment]
> Hi Leonard, the change is partially merged now. I kept some things out that I wasn't quite sure if they are needed. I thought it's best to get everything else in first and then we can take it from there. It doesn't mean that I object to the other changes.
> 
> I will be out of office until next week, but happy to review more on Monday.

Sounds good to me. Do you have a list of which pieces were held out? (I don't see any differences between what's merged in and my `windows_tfrt` branch)

**Comment by chsigg** [comment]
I only remember the struct -> class changes. I'm terrible with git, but you should be able to rebase your branch from master to see what's missing?
